File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        
        final var result = new int[nums.length];

        var prevValue = 1;
        var prevProduct = 1;

        for (int i = 0; i < nums.length; i++) {

            final var value = nums[i];
            result[i] = prevProduct * prevValue;

            prevValue = value;
            prevProduct = result[i];
        }

        prevProduct = 1;

        for (int i = nums.length - 1; i >= 0; i--) {

            final var value = nums[i];
            result[i] *= prevProduct;
            prevProduct *= value;
        }

        return result;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left_side_product = 1
        prev_value = 1
        result = [0] * len(nums)

        for i in range(len(nums)):
            left_side_product *= prev_value
            result[i] = left_side_product
            prev_value = nums[i]
        
        right_side_product = 1

        for i in reversed(range(len(nums))):
            result[i] *= right_side_product
            right_side_product *= nums[i]

        return result